consensus/hotstuff/timeoutaggregator/timeout_aggregator.go

```diff
@@ -0,0 +1,231 @@
+	})
+
+	aggregator.ComponentManager = componentBuilder.Build()
+	return aggregator, nil
+}
+
+// queuedTimeoutsProcessingLoop is the event loop which waits for notification about pending work
+// and as soon as there is some it triggers processing.
+// All errors are propagated to SignalerContext
+func (t *TimeoutAggregator) queuedTimeoutsProcessingLoop(ctx irrecoverable.SignalerContext) {
+	notifier := t.queuedTimeoutsNotifier.Channel()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-notifier:
+			err := t.processQueuedTimeoutEvents(ctx)
+			if err != nil {
+				ctx.Throw(fmt.Errorf("internal error processing queued timeout events: %w", err))
+				return
+			}
+		}
+	}
+}
+
+// processQueuedTimeoutEvents is a function which dispatches previously queued timeouts on worker thread
+// This function is called whenever we have queued timeouts ready to be dispatched.
+// No errors are expected during normal operations.
+func (t *TimeoutAggregator) processQueuedTimeoutEvents(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		msg, ok := t.queuedTimeouts.Pop()
+		if ok {
+			timeoutObject := msg.(*model.TimeoutObject)
+			err := t.processQueuedTimeout(timeoutObject)
+			if err != nil {
+				return fmt.Errorf("could not process pending TO %v: %w", timeoutObject.ID(), err)
+			}
+
+			t.log.Info().
+				Uint64("view", timeoutObject.View).
+				Str("timeout_id", timeoutObject.ID().String()).
+				Msg("TO has been processed successfully")
+
+			continue
+		}
+
+		// when there is no more messages in the queue, back to the loop to wait
+		// for the next incoming message to arrive.
+		return nil
+	}
+}
+
+// processQueuedTimeout performs actual processing of queued timeouts, this method is called from multiple
+// concurrent goroutines.
+func (t *TimeoutAggregator) processQueuedTimeout(timeoutObject *model.TimeoutObject) error {
+	// TODO: replace this check by a specific function to query if there is epoch by view
+	_, err := t.committee.LeaderForView(timeoutObject.View)
+	if err != nil {
+		// ignore TO if we don't have information for epoch
+		if errors.Is(err, model.ErrViewForUnknownEpoch) {
+			return nil
+		}
+		return fmt.Errorf("unknown error when querying epoch state for view %d: %w", timeoutObject.View, err)
+	}
+
+	collector, created, err := t.collectors.GetOrCreateCollector(timeoutObject.View)
+	if err != nil {
+		// ignore if our routine is outdated and some other one has pruned collectors
+		if mempool.IsDecreasingPruningHeightError(err) {
+			return nil
+		}
+		return fmt.Errorf("could not get collector for view %d: %w",
+			timeoutObject.View, err)
+	}
+	if created {
+		t.log.Info().Uint64("view", timeoutObject.View).Msg("timeouts collector is created by processing TO")
+	}
+
+	err = collector.AddTimeout(timeoutObject)
+	if err != nil {
+		if model.IsDoubleTimeoutError(err) {
+			doubleTimeoutErr := err.(model.DoubleTimeoutError)
```

Comment:
## ⚠️ ❓ 
`DoubleTimeoutError` has already a convenience function `AsDoubleTimeoutError`, which includes the type cast. This is important, because `err` might be a _wrapped_ `DoubleTimeoutError` in which cast the case would panic (I think?)
```suggestion
		if doubleTimeoutErr, is := model.AsDoubleTimeoutError(err); is {
```